backend/src/controllers/upload.ts

```diff
@@ -1,67 +1,82 @@
-  storage = new Storage({keyFilename: fullConfig.googleSvcAcctKey});
+  storage = new Storage({ keyFilename: fullConfig.googleSvcAcctKey });
 }
 
-const storageBucketName = fullConfig.storageBucketName
+const storageBucketName = fullConfig.storageBucketName;
+
+const UploadRequestQuerySchema = z.object({
+  originalFileName: z.string(),
+  base64File: z
+    .string()
+    .max(MAX_UPLOAD_MB * 1024 * 1024, "File size too large"),
+});
 
 class UploadController {
   async upload(req: express.Request, res: express.Response) {
-    // check first before proceeding
-    if (req.body.originalFilename === undefined || req.body.originalFilename === "") {
-      const result = {"status": "failure", "message": "Invalid original filename"};
-      return res.end(JSON.stringify(result));
-    }
+    try {
+      const input = UploadRequestQuerySchema.parse(req.body);
 
-    const MAX_UPLOAD_MB = 10;
-    if (req.body.base64File.length / 4 * 3 > MAX_UPLOAD_MB * 1024 * 1024) {
-      const result = {"status": "failure", "message": "File size too large"};
-      return res.end(JSON.stringify(result));
-    }
+      const mimeType = mime.lookup(input.originalFileName);
+      const buffer = Buffer.from(input.base64File, "base64");
 
-    const bufferStream = new stream.PassThrough();
-    bufferStream.end(Buffer.from(req.body.base64File, 'base64'));
-
-    //Define bucket.
-    const myBucket = storage.bucket(storageBucketName);
-    //Define file & file name.
-    const newFileName = makeRandomName();
-    const file = myBucket.file(newFileName);
-    //Pipe the 'bufferStream' into a 'file.createWriteStream' method.
-    bufferStream.pipe(file.createWriteStream())
-      .on('error', function(err) {
-        console.log("Error during file upload: " + err.toString())
-        console.log(err);
-        const result = {"status": "failure", "message": "Failed to upload file"};
-        return res.end(JSON.stringify(result));
-      })
-      .on('finish', function() {
-        // create object in database
-        const data = {
-          userId: req.user.id,
-          originalFilename: req.body.originalFilename,
-          filename: newFileName,
-        };
-        Upload.create(data).then((newUpload: Upload) => {
-          if (!newUpload) {
-            const result = {"status": "failure", "message": "Failed to create upload reference"};
-            return res.end(JSON.stringify(result));
-          }
-          if (newUpload) {
-            const result = {"status": "success", "message": "You have successfully uploaded the file", "payload": newUpload};
-            return res.end(JSON.stringify(result));
-          }
+      //Define bucket.
+      const uploadBucket = storage.bucket(storageBucketName);
+      //Define file & file name.
+      const newFileName = makeRandomName();
+      const file = uploadBucket.file(newFileName);
+
+      try {
+        await file.save(buffer, {
+          contentType: mimeType ? mimeType : undefined,
```

Comment:
should we reject the upload if a funny file extension is used?

Reply:
based on the OWASP recommendation we should whitelist allowed file types, and try to keep that list as small as possible, so yes, we should validate for those

we can reject files that don't have extensions too? if enough users report that they have files or pictures without extensions (?!) then we can look at this again

Reply:
Yup. We should whitelist. Most users won't even know how to view images without extensions. I assume most uploaders will be like most users, so we can reject missing file extensions and instruct them to use an appropriate extension if the edge use case appears.